src/components/PostDetails.tsx

```diff
@@ -1,117 +1,218 @@
 import { Loader } from './Loader';
 import { NewCommentForm } from './NewCommentForm';
 
-export const PostDetails: React.FC = () => {
+import { Post } from '../types/Post';
+import { User } from '../types/User';
+import { Comment, CommentData } from '../types/Comment';
+import { Errors } from '../types/Errors';
+
+import { getComments, postComment, deleteComment } from '../api/users';
+
+type Props = {
+  post: Post | null;
+  user: User | null;
+  isFormShown: boolean;
+  typeOfError: Errors;
+  setTypeOfError: (data: Errors) => void;
+  showForm: () => void;
+};
+
+export const PostDetails: React.FC<Props> = ({
+  post,
+  user,
+  isFormShown,
+  showForm,
+  typeOfError,
+  setTypeOfError,
+}) => {
+  const [commentsList, setCommentsList] = useState<Comment[] | null>(null);
+  const [areCommentsLoading, setCommentsLoading] = useState(false);
+  const [isCommentAdding, setCommentAdding] = useState(false);
+  const [
+    removingCommentsId, setRemovingCommentsId,
+  ] = useState<number | null>(null);
+  const [newCommentsData, setNewCommentsData] = useState<CommentData>(
+    {
+      postId: post?.id || null,
+      name: '',
+      email: '',
+      body: '',
+    },
+  );
+
+  async function loadComments() {
+    setTypeOfError(Errors.None);
+    try {
+      setCommentsLoading(true);
+      const commentsFromServer = await getComments(user?.id);
+
+      setCommentsList(commentsFromServer);
+    } catch (error) {
+      setTypeOfError(Errors.Comments);
+    } finally {
+      setCommentsLoading(false);
+    }
+  }
+
+  const addComment = async () => {
+    setCommentAdding(false);
+    setTypeOfError(Errors.None);
+
+    try {
+      setCommentAdding(true);
+
+      const comment = await postComment(newCommentsData);
+
+      setCommentsList(currComments => {
+        if (currComments) {
+          return [...currComments, comment];
+        }
+
+        return [comment];
+      });
+    } catch (error) {
+      setTypeOfError(Errors.Adding);
+    } finally {
+      setCommentAdding(false);
+    }
+  };
+
+  const removeComment = async (id: number) => {
+    setTypeOfError(Errors.None);
+    setRemovingCommentsId(null);
+    try {
+      await deleteComment(id);
+
+      setCommentsList(comments => {
+        return comments?.filter(comment => comment.id !== id) || null;
+      });
+    } catch (error) {
+      setTypeOfError(Errors.Deleting);
+      setRemovingCommentsId(id);
+    }
+  };
+
+  useEffect(() => {
+    loadComments();
+    setNewCommentsData(
+      {
+        postId: post?.id || null,
+        name: '',
+        email: '',
+        body: '',
+      },
+    );
+  }, [post?.id]);
+
+  useEffect(() => {
+    if (typeOfError === Errors.Adding || typeOfError === Errors.Deleting) {
+      const timer = window.setTimeout(() => {
+        setTypeOfError(Errors.None);
+        setRemovingCommentsId(null);
+      }, 3000);
+
+      return () => {
+        window.clearTimeout(timer);
+      };
+    }
+
+    return () => {};
```

Comment:
You can remove the empty return statement at the end of this useEffect, as it's not needed.

Reply:
![image](https://github.com/mate-academy/react_dynamic-list-of-posts/assets/122163374/ce661eee-c6c8-4ce2-96ce-ca8e95f923c1)

Reply:
I don't have an error with this code

```
    useEffect(() => {
        if (typeOfError === Errors.Adding || typeOfError === Errors.Deleting) {
            const timer = window.setTimeout(() => {
                setTypeOfError(Errors.None);
                setRemovingCommentsId(null);
            }, 3000);

            return () => {
                window.clearTimeout(timer);
            };
        }
    }, [typeOfError, setTypeOfError, setRemovingCommentsId]);
```